File: server/lib/hyperarchy/models/organization.rb
class Organization < Monarch::Model::Record
  column :name, :string
  column :description, :string, :default => ""
  column :members_can_invite, :boolean, :default => false
  column :dismissed_welcome_guide, :boolean, :default => false
  column :use_ssl, :boolean, :default => true
  column :election_count, :integer, :default => 0
  column :member_count, :integer, :default => 0
  column :created_at, :datetime
  column :updated_at, :datetime
  column :social, :boolean, :default => false
  column :privacy, :string, :default => "private"
  column :invitation_code, :string

  has_many :elections, :order_by => "score desc"
  has_many :memberships

  relates_to_many :members do
    memberships.join_through(User)
  end

  attr_accessor :suppress_membership_creation

  def self.social
    find(:social => true)
  end

  def guest
    members.find(:guest => true)
  end

  def can_create?
    current_user && !current_user.guest?
  end

  def can_update_or_destroy?
    current_user.admin? || has_owner?(current_user)
  end
  alias can_update? can_update_or_destroy?
  alias can_destroy? can_update_or_destroy?

  def write_blacklist
    [:social]
  end

  def before_create
    self.invitation_code = SecureRandom.hex(8)
  end

  def after_create
    special_guest = User.create_guest(id)
    memberships.create(:user => special_guest, :pending => false)
    memberships.create(:user => current_user, :role => "owner", :pending => false) unless suppress_membership_creation
  end

  def has_member?(user)
    !memberships.find(:user_id => user.id).nil?
  end

  def current_user_is_member?
    has_member?(current_user)
  end

  def ensure_current_user_is_member
    raise Monarch::Unauthorized unless current_user
    return if current_user_is_member?
    raise Monarch::Unauthorized unless public?
    memberships.create!(:user => current_user, :pending => false)
  end

  def current_user_can_create_items?
    (public? && !current_user.guest?) || has_member?(current_user)
  end

  def has_owner?(user)
    !memberships.find(:user_id => user.id, :role => "owner").nil?
  end

  def current_user_is_owner?
    has_owner?(current_user)
  end

  def allow_subscription?(user)
    if public_readable?
      true
    else
      user.admin? || has_member?(user)
    end
  end

  def public?
    privacy == "public"
  end

  def public_readable?
    privacy == "public" || privacy == "read_only"
  end

  def private?
    privacy == "private"
  end

  def organization_ids
    [id]
  end
end

File: app/assets/javascripts/models/organization.js
_.constructor("Organization", Model.Record, {
  constructorProperties: {
    initialize: function() {
      this.columns({
        name: "string",
        description: "string",
        membersCanInvite: "boolean",
        electionCount: 'integer',
        memberCount: 'integer',
        useSsl: 'boolean',
        social: 'boolean',
        privacy: 'string',
        membershipCode: 'string'
      });

      this.hasMany("elections");
      this.relatesToMany("candidates", function() {
        return this.elections().joinThrough(Candidate);
      });
      this.relatesToMany("votes", function() {
        return this.elections().joinThrough(Vote);
      });

      this.hasMany("memberships", {orderBy: ["firstName asc", "emailAddress asc"]});
      this.relatesToMany("members", function() {
        return this.memberships().joinThrough(User);
      });
    },

    findSocial: function() {
      return this.find({social: true});
    }
  },

  afterInitialize: function() {
    this.numElectionsFetched = 0;
  },

  fetchMoreElections: function() {
    if (this.fetchInProgressFuture) return this.fetchInProgressFuture;

    if (this.numElectionsFetched >= this.electionCount()) {
      var promise = new Monarch.Promise();
      promise.triggerSuccess();
      return promise;
    }

    // if we already fetched some, fetch 8 positions back to account for unseen swapping at the fringe
    var offset, limit;
    if (this.numElectionsFetched > 0) {
      offset = this.numElectionsFetched - 8;
      limit = 24;
    } else {
      offset = 0;
      limit = 16;
    }

    var promise = $.ajax({
      url: "/elections",
      data: {
        organization_id: this.id(),
        offset: offset,
        limit: limit
      },
      dataType: 'records'
    });

    this.fetchInProgressFuture = promise;
    promise.success(this.bind(function() {
      delete this.fetchInProgressFuture;
      this.numElectionsFetched += 16;
    }));

    return promise;
  },

  membershipForUser: function(user) {
    return this.memberships().find({userId: user.id()});
  },

  membershipForCurrentUser: function() {
    return this.membershipForUser(Application.currentUser());
  },

  currentUserIsMember: function() {
    return this.membershipForCurrentUser() != null;
  },

  currentUserIsOwner: function() {
    var currentUserMembership = this.memberships().find({userId: Application.currentUserId()});
    if (!currentUserMembership) return false;
    return currentUserMembership.role() === "owner";
  },

  ensureCurrentUserCanParticipate: function() {
    var future = new Monarch.Http.AjaxFuture();
    if (!this.isPublic() && !this.currentUserIsMember()) {
      Application.layout.mustBeMemberMessage.show();
      future.triggerFailure();
    } else if (Application.currentUser().guest()) {
      Application.layout.signupPrompt.future = future;
      Application.layout.signupPrompt.showSignupForm();
      Application.layout.signupPrompt.show()
    } else {
      future.triggerSuccess();
    }
    return future;
  },

  currentUserCanParticipate: function() {
    return !Application.currentUser().guest() && (this.isPublic() || this.currentUserIsMember());
  },

  currentUserCanEdit: function() {
    return Application.currentUser().admin() || this.currentUserIsOwner();
  },

  currentUserCanInvite: function() {
    return this.currentUserIsOwner() || (this.currentUserIsMember() && this.membersCanInvite());
  },

  isPublic: function() {
    return this.privacy() === "public";
  },

  hasNonAdminQuestions: function() {
    return !this.elections().where(Election.creatorId.neq(1)).empty();
  },

  url: function() {
    return "/organizations/" + this.id();
  },

  settingsUrl: function() {
    return this.url() + "/settings";
  },

  newElectionUrl: function() {
    return this.url() + "/elections/new";
  },

  secretUrl: function() {
    return 'https://' + Application.HTTP_HOST + "/access/" + this.id() + "/" + this.membershipCode();
  }
});
